README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active) [![Linux Build Status](https://travis-ci.org/mdsumner/gris.svg?branch=master)](https://travis-ci.org/mdsumner/gris) [![Windows Build status](https://ci.appveyor.com/api/projects/status/github/mdsumner/gris?svg=true)](https://ci.appveyor.com/project/mdsumner/gris) [![](http://www.r-pkg.org/badges/version/gris)](http://cran.rstudio.com/web/packages/gris/index.html) [![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/gris)](http://cran.r-project.org/web/packages/gris/index.html) [![Coverage Status](https://img.shields.io/codecov/c/github/mdsumner/gris/master.svg)](https://codecov.io/github/mdsumner/gris?branch=master)

gris
====

A database table model for storing geometry in R.

The gris package provides a relational geometry/topology model for spatial data in R. This is inspired by data models used in the commercial products Manifold System GIS and Eonfusion. The main aspirations are

1.  remove the X/Y limitation on vertex attributes for points, lines, surfaces (and polygons)
2.  allow multiple topology types in individual layers
3.  provide a flexible basis for conversion between other formats.
4.  (a similar generalization for raster data)

Installation
------------

``` r
devtools::install_github("mdsumner/gris")
```

Quick example
-------------

Convert `SpatialPolygonsDataFrame` to `gris`. This shows the basics, that we can store Spatial-structures without loss of information.

``` r
library(gris)
library(maptools);data(wrld_simpl)
#> Loading required package: sp
#> Checking rgeos availability: FALSE
#>      Note: when rgeos is not available, polygon geometry     computations in maptools depend on gpclib,
#>      which has a restricted licence. It is disabled by default;
#>      to enable gpclib, type gpclibPermit()
gg <- gris(wrld_simpl)
plot(gg)
```

![](figures/README-unnamed-chunk-3-1.png)

``` r
str(gg)
#> List of 5
#>  $ o     :Classes 'tbl_df', 'tbl' and 'data.frame':  246 obs. of  12 variables:
#>   ..$ FIPS     : Factor w/ 244 levels "","AC","AE","AF",..: 2 5 6 7 8 10 11 12 13 17 ...
#>   ..$ ISO2     : Factor w/ 246 levels "AD","AE","AF",..: 4 61 17 6 7 9 12 11 14 24 ...
#>   ..$ ISO3     : Factor w/ 246 levels "ABW","AFG","AGO",..: 15 64 18 6 11 3 12 10 16 25 ...
#>   ..$ UN       : int [1:246] 28 12 31 8 51 24 16 32 36 48 ...
#>   ..$ NAME     : Factor w/ 246 levels "Aaland Islands",..: 10 4 16 3 12 7 5 11 14 18 ...
#>   ..$ AREA     : int [1:246] 44 238174 8260 2740 2820 124670 20 273669 768230 71 ...
#>   ..$ POP2005  : num [1:246] 83039 32854159 8352021 3153731 3017661 ...
#>   ..$ REGION   : int [1:246] 19 2 142 150 142 2 9 19 9 142 ...
#>   ..$ SUBREGION: int [1:246] 29 15 145 39 145 17 61 5 53 145 ...
#>   ..$ LON      : num [1:246] -61.78 2.63 47.4 20.07 44.56 ...
#>   ..$ LAT      : num [1:246] 17.1 28.2 40.4 41.1 40.5 ...
#>   ..$ .ob0     : int [1:246] 1 2 3 4 5 6 7 8 9 10 ...
#>  $ b     :Classes 'tbl_df', 'tbl' and 'data.frame':  3768 obs. of  3 variables:
#>   ..$ .br0: int [1:3768] 1 2 3 4 5 6 7 8 9 10 ...
#>   ..$ .ob0: int [1:3768] 1 1 2 3 3 3 3 3 4 5 ...
#>   ..$ .h0 : int [1:3768] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ bXv   :Classes 'tbl_df', 'tbl' and 'data.frame':  22497 obs. of  3 variables:
#>   ..$ .vx0     : int [1:22497] 5589 5620 5605 5596 5611 5613 16101 17581 18103 18790 ...
#>   ..$ .br0     : int [1:22497] 1 1 1 2 2 2 3 3 3 3 ...
#>   ..$ .br_order: int [1:22497] 1 2 3 1 2 3 1 2 3 4 ...
#>  $ v     :Classes 'tbl_df', 'tbl' and 'data.frame':  21165 obs. of  3 variables:
#>   ..$ x   : num [1:21165] -61.7 -61.9 -61.8 -61.7 -61.9 ...
#>   ..$ y   : num [1:21165] 17 17.1 17.2 17.6 17.6 ...
#>   ..$ .vx0: int [1:21165] 5589 5620 5605 5596 5611 5613 16101 17581 18103 18790 ...
#>  $ georef:List of 1
#>   ..$ proj4: chr " +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0"
#>   ..- attr(*, "class")= chr [1:2] "georef" "list"
#>  - attr(*, "class")= chr [1:2] "gris" "list"
```

Convert `gris` to rgl and plot in XYZ.

``` r
gt <- triangulate(gg)
#> Joining, by = ".br0"
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
str(gt)
#> List of 7
#>  $ o     :Classes 'tbl_df', 'tbl' and 'data.frame':  246 obs. of  12 variables:
#>   ..$ FIPS     : Factor w/ 244 levels "","AC","AE","AF",..: 2 5 6 7 8 10 11 12 13 17 ...
#>   ..$ ISO2     : Factor w/ 246 levels "AD","AE","AF",..: 4 61 17 6 7 9 12 11 14 24 ...
#>   ..$ ISO3     : Factor w/ 246 levels "ABW","AFG","AGO",..: 15 64 18 6 11 3 12 10 16 25 ...
#>   ..$ UN       : int [1:246] 28 12 31 8 51 24 16 32 36 48 ...
#>   ..$ NAME     : Factor w/ 246 levels "Aaland Islands",..: 10 4 16 3 12 7 5 11 14 18 ...
#>   ..$ AREA     : int [1:246] 44 238174 8260 2740 2820 124670 20 273669 768230 71 ...
#>   ..$ POP2005  : num [1:246] 83039 32854159 8352021 3153731 3017661 ...
#>   ..$ REGION   : int [1:246] 19 2 142 150 142 2 9 19 9 142 ...
#>   ..$ SUBREGION: int [1:246] 29 15 145 39 145 17 61 5 53 145 ...
#>   ..$ LON      : num [1:246] -61.78 2.63 47.4 20.07 44.56 ...
#>   ..$ LAT      : num [1:246] 17.1 28.2 40.4 41.1 40.5 ...
#>   ..$ .ob0     : int [1:246] 1 2 3 4 5 6 7 8 9 10 ...
#>  $ b     :Classes 'tbl_df', 'tbl' and 'data.frame':  3768 obs. of  3 variables:
#>   ..$ .br0: int [1:3768] 1 2 3 4 5 6 7 8 9 10 ...
#>   ..$ .ob0: int [1:3768] 1 1 2 3 3 3 3 3 4 5 ...
#>   ..$ .h0 : int [1:3768] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ bXv   :Classes 'tbl_df', 'tbl' and 'data.frame':  22497 obs. of  3 variables:
#>   ..$ .vx0     : int [1:22497] 5589 5620 5605 5596 5611 5613 16101 17581 18103 18790 ...
#>   ..$ .br0     : int [1:22497] 1 1 1 2 2 2 3 3 3 3 ...
#>   ..$ .br_order: int [1:22497] 1 2 3 1 2 3 1 2 3 4 ...
#>  $ v     :Classes 'tbl_df', 'tbl' and 'data.frame':  21165 obs. of  3 variables:
#>   ..$ x   : num [1:21165] -61.7 -61.9 -61.8 -61.7 -61.9 ...
#>   ..$ y   : num [1:21165] 17 17.1 17.2 17.6 17.6 ...
#>   ..$ .vx0: int [1:21165] 5589 5620 5605 5596 5611 5613 16101 17581 18103 18790 ...
#>  $ georef:List of 1
#>   ..$ proj4: chr " +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0"
#>   ..- attr(*, "class")= chr [1:2] "georef" "list"
#>  $ tXv   :Classes 'tbl_df', 'tbl' and 'data.frame':  14966 obs. of  5 variables:
#>   ..$ .vx1: int [1:14966] 5589 5596 17800 17781 17791 17802 17778 17787 11068 2608 ...
#>   ..$ .vx2: int [1:14966] 5620 5611 17801 17778 17790 17801 17781 17790 11062 2610 ...
#>   ..$ .vx3: int [1:14966] 5605 5613 17802 17775 17787 17800 17775 17791 11077 2613 ...
#>   ..$ .tr0: int [1:14966] 1 2 3 4 5 6 7 8 9 10 ...
#>   ..$ .br0: int [1:14966] 1 2 6 7 8 11 12 13 16 17 ...
#>  $ oXt   :Classes 'tbl_df', 'tbl' and 'data.frame':  14966 obs. of  2 variables:
#>   ..$ .tr0: int [1:14966] 1 2 3 4 5 6 7 8 9 10 ...
#>   ..$ .ob0: int [1:14966] 1 1 3 3 3 5 5 5 6 7 ...
#>  - attr(*, "class")= chr [1:2] "gris" "list"
gris::plot3d(gt)
rgl::rgl.snapshot("figures/xyz.png")
```

![xyz](figures/xyz.png)

What we have
------------

-   build relational table structures representing vector GIS and conversion to network models and visualization models
-   `Branches` and `Primitives` model to link vertices and objects, including coexistence of both models on one structure
-   simplistic raster conversion to OpenGL structures

What we need
------------

-   Primitives model is currently only triangles, this needs generalization for line segments, for representation and plotting, and some resolution of whether "wide" (v1, v2, v3) or "long" (v0, p0) format is preferable (the former is like the structural form in OpenGL, the latter more relational)
-   need point-in-poly tests independent of sp for determining holes and filtering out triangles (currently using geometry::tsearch)
-   invalidation, i.e. if we manipulate the mesh of triangles either the branches get removed or re-calculated
-   need to fix how gris objects are created, with a proper set of constructor methods, not that g\[i, \] duplicates .ob0 for repeated i values - but maybe it's nonsense to allow repeats anyway?
-   control over coordinate system (crs) metadata (PROJ.4 is not enough, need to be flexible to change, proj4 rather than rgdal etc.)
-   ability to explode and union objects arbitrarily, and general access to lower level tools (tools in sp/rgeos should be decoupled from the Spatial classes, for example)
-   system of naming for vertices in order to generalize away from x/y
-   analogue to the vector case for raster data, to provide n-dimensional curvilinear grids with rectilinear and affine dimensions as special cases

### Storage of triangles and branches together

This is a valuable feature to have, but it's not yet obvious what is the best way to go. Complicating factors are:

-   storage of end coordinate for polygons (gris doesn't do this, though the method used keeps them at first so this needs review)
-   object triangulation for minimum area or angle - this needs vertices inserted so updating IDs per object is more complicated
-   normalization of vertices by branch is possible by triangulating twice, first the holes and then the rest, or by matching original IDs as the vertices are re-collated
-   it may make sense to build the triangular mesh, with holes removed, then find cycles and rebuild the branches - but this has to happen on a per object basis, unless we somehow carry the line segment indexes through

### Links to source data

Triangle meshes can be combined and re-triangulated fairly simply, this means we can work with the union of two or more layers (the Identity intersection etc.). We need a system of recording these links to inputs.

Topology
========

Currently in gris, traditional GIS-like objects are decomposed into a set of relational tables. These are Vertices and Objects, and these two tables may be linked via the *Branches model* with intermediate tables `Vertex-Link-Branches` (bXv) and `Branches` (b), or via the *Primitives model* with intermediate tables `Vertex-Link-Triangles` (tXv) and `Object-Link-Triangles` (oXt).

Storing these data in tables is similar to the *ggplot2 model* where a GIS object is converted to a single table with `fortify`, but this duplicates object and branch information by copying them onto every vertex. This is straightforward, but wasteful and does not provide a logical structure for entitities, these are implicit and must be derived by uniqueness tests and so on. In *gris* the `Objects` table is analogous to the *attributes* table in a GIS, each row stores data on an object which may be topologically a (multi) polygon, line, or point. The Branches table allows for multiple vertices on a a single point, multiple connected sets of line segments (linear strings) for a single line, and multiple "rings" for a complex polygon that may contain "holes" or include multiple separate "islands".

Vertex-Link-Branches is necessary so that the vertex table can be normalized, i.e. remove duplicates in x/y (or more generally any combination of attributes). It may be worth having a Branches-Link-Objects table (and triangles-link-objects) for further normalization of complex layers, but I ignore it for now.

There are three steps in converting a traditional polygon into a surface: 1) decompose the polygon into a set of vertices with a two-element index specifying every boundary segment, a planar-straight-line-graph (the vertices must also be "normalized" to remove any duplicates) 2) triangulate with Triangle's\*\* constrained Delaunay algorithm (optionally with a maximum triangle area constraint) 3) calculate centroids of every triangle and filter those that compose a "hole" in the original complex polygon.

Gris converts to the Branches topology by default, but can also use Shewchuk's Triangle to decompose each object to Constrained Delaunay triangular primitives and simply insert tables to link the objects and vertices by this alternative. It would be helpful to include other triangulation methods, like the unconstrained algorithms in *deldir*, *geometry* and *spatstat*, and the ear-clipping (random or deterministic) algorithms in *rgl*. I'll build in a connection to Manifold's constrained algorithm via the *manifoldr* package.

Other important comparisons must be done with topojson and D3.

There is a wrinkle in keeping the relation between objects and primitives in that it may be necessary to perform the triangulation on an object-by-object basis. This complicates the approach since the Triangle vertex index is *structural* rather than relational, but it's working and these notes may need revision:

1.  assume that the addition of Steiner points to the triangulation is always appended to the input vertices
2.  always maintain the entire set of input vertices, but filter the PSLG segments for each triangulation (in fact this might mean doing the entire triangulation upfront, then iterating again through every object in turn)

\*\*Triangle currently provides the only easily accessible constrained Delaunay triangulation algorithm. CGAL does provide boundary constraints, but is pretty hard to use and maybe?? doesn't easily allow maximum triangle area. Manifold does boundary constraints, but not maximum area and is affordable, but not open. Spatstat, deldir, GEOS/rgeos and geometry packages all have Delaunay but not with constraints. GDAL (i.e. Even Roualt) is developing Delaunay algorithms due in early 2016. PostGIS has GEOS capability, maybe something more? Eonfusion had constrained triangulations but not further subdivision for maximum area of triangles, and the decomposition to primitives was always done upfront in the native crs which was not necessarily sensible.

Related
=======

[Notes](notes_links.md)


NEWS.md

# gris 0.1.0

* replace llh2xyz with geocentric, driving proj4 'geocent'

* removed quadmeshFromRaster (see quadmesh package on CRAN)

* many updates to align to spbabel

* updated dplyr::distinct usage


# gris 0.0.6.9001

* Added a `NEWS.md` file to track changes to the package.

* renamed `bgl` to `quadmeshFromRaster`

* added detail to help topic `gris-package`

* cleaned up some functions not needed to be exported 

R/gris.R
## edge  pairs from sequential indexes for polygons
prs1 <- function(x) {
  x1 <- cbind(head(x,-1), tail(x,-1))
  rbind(x1, c(x1[length(x1)], x1[1]))
}


## simpler version of mkpslg, to only make edges from branches

mkedges <- function(x) {
  v <- x$v
  bXv <- x$bXv
  v$remap <- seq(nrow(v))
  bXv$vertex_ <- v$remap[match(bXv$vertex_, v$vertex_)]
  E <- do.call(rbind, lapply(split(
    bXv$vertex_, bXv$branch_
  ), prs1))
  ## this is structural for now, needs to store the vertex index
  x$e <- data_frame(s0 = E[,1], s1 = E[,2])
  x
}
  
  

# Planar Straight Line Graph
# 
# Constrained triangulation requires a Planar Straight Line Graph consisting of
# unique vertices and an index matrix for each line segment edge. This function builds that object.
# @param x  gris object
# 
# @return \code{\link[RTriangle]{pslg}} Planar Straight Line Graph object
# @importFrom RTriangle pslg
# @importFrom dplyr select %>%
# @export
mkpslg <- function(x) {
  ## remap vertices
  x$v$remap <- seq(nrow(x$v))
  x$bXv$vertex_ <- x$v$remap[match(x$bXv$vertex_, x$v$vertex_)]
  RTriangle::pslg(
    P = x$v %>% dplyr::select_(.dots = c("x_", "y_")) %>% as.matrix(),
    S = do.call(rbind, lapply(split(
      x$bXv$vertex_, x$bXv$branch_
    ), prs1))
  )
  
}


#' gris
#'
#' @param x Spatial* object
#' @param ... unused
#'
#' @return gris
#' @export
#' @section Methods:
#'
#' \code{gris} implements these base methods:
#'
#' \describe{
#' \item{plot}{Plots the object according to its topology type, or by a chosen \code{type}. This
#' method follows base plotting with types "p" points, "l" lines, and adds "pp" for polygons}
#' \item{print}{Prints the object and vertices tables with a summary of the topology linkages}
#' \item{\code{[}}{Subset as an object table on attributes, propagates down through the geometry
#' links to return a subsetted gris object}
#' }
gris <- function(x, ...) {
  UseMethod("gris")
}

gris.full <- function(o,  b, bXv, v, georef = NULL) {
  v$vertex_ <- as.integer(v$vertex_)
  bXv$vertex_ <- as.integer(bXv$vertex_)
  bXv$order_ <- as.integer(bXv$order_)
  bXv$branch_ <- as.integer(bXv$branch_)
  b$branch_ <- as.integer(b$branch_)
  if ("island_" %in% names(b)) b$island_ <- as.integer(b$island_)
  b$object_ <- as.integer(b$object_)
  o$object_ <- as.integer(o$object_)
  x <- list(
    o = o,  b = b, bXv = bXv, v = v, 
    georef = georef
  )
  class(x) <- c("gris", "list")
  x
}

## vectors of (possibly named) arguments in ... or in a list
#' @export
#' @importFrom tibble as_tibble
gris.default <- function(..., topotype = "p") {
  x <- list(...)
  as_tibble(setNames(x, buildnames(x)))
}

# build names as required, but preserve any that are input
# generated names may clobber later ones that are specified, cest la vie
buildnames <- function(x) {
  nams <- names(x)
  if (is.null(nams)) {
    nams <-  dfn(x)
  }
  nonames <- nchar(nams) < 1
  if (any(nonames)) {
    nams[nonames] <- dfn(x)[nonames]
  }
  make.unique(nams, sep = "")
}


# generate x, y, z, t and then as many more as required appending integers to the alphabet
dfn <- function(x) {
  ind <- c(24, 25, 26, 20)
  first <-
    rep_len(c(letters[ind], letters[-ind], letters), length.out = length(x))
  make.unique(sprintf("%s_", first, sep = ""))
}

#' @rdname gris
#' @param i indices for extract
#' @param j indices for extract
#' @param drop ignored
#' @export
#' @importFrom dplyr semi_join
`[.gris` <- function (x, i, j, drop = FALSE) {
  do_tri <- "tXv" %in% names(x)  
  o <- x$o[i,j,drop = drop]

  b <- x$b %>% semi_join(o, by = "object_")
  bXv <- x$bXv %>% semi_join(b, by = "branch_")
  v <- x$v %>% semi_join(bXv, by = "vertex_")
  if (do_tri) {
    oXt <- x$oXt %>% semi_join(o, by = "object_")
    tXv <- x$tXv %>% semi_join(oXt, by = ".tr0")
  #  v <- bind_rows(x$v %>% semi_join(tXv, c("vertex_" = ".vx1")),
  #                  x$v %>% semi_join(tXv, c("vertex_" = ".vx2")),
  #                  x$v %>% semi_join(tXv, c("vertex_" = ".vx3"))) %>%
  #    distinct_("vertex_")

  } else {
    v <- x$v %>% semi_join(bXv, by = "vertex_")
  }
  #gris.full(o, oXb, b, bXv, v)
  x <- gris.full(o,  b, bXv, v)
  if (do_tri) {
    x$tXv <- tXv 
   x$oXt <- oXt
  }
  x
}

#' @rdname gris
#' @param width width
#' @param n n
#' @param y y
#' @export 
print.gris <- function(x, ..., n = NULL, width = NULL) {
  cat("gris object", "\n", sep = "")
  print(x$o)
  cat("\n")
  cat("gris topology ", topotype(x), "\n", sep = "")
  cat("\n")
  cat("gris vertices", "\n", sep = "")
  print(x$v)
  cat("\n")
  invisible(x)
}

#' @rdname gris
#' @export
#' @importFrom dplyr do ungroup select
plot.gris <- function(x, y,  triangles  = FALSE, ...) {
  xrange <- range(x$v$x_, na.rm = TRUE)
  yrange <- range(x$v$y_, na.rm = TRUE)
  largs <- list(...)
  if (is.null(largs$type)) {
    type <- "pp"  ## default to polygon for now
  }
  rule <- "evenodd"
  if (!is.null(largs$rule)) {
    if (largs$type != "pp") warning("argument 'rule' ignored for non polygon plot")
    rule <- largs$rule 
  }
  # type, rule
  if (is.null(largs$add) || !largs$add) {
    #otype <- largs$type
    largs$type <- "n"
    largs$add <- NULL
    largs$x <- xrange
    largs$y <- yrange
    do.call(plot, largs)
    largs$type <- NULL
  }
  if(triangles) {
    plotT(x ,border = "black")
    return(invisible(x))
  }
  if (type == "pp") largs$rule <- rule
  uoid <- unique(x$o$object_)
  if (is.null(largs$col))
    largs$col <- sample(grey(seq_along(uoid) / (length(uoid) + 1)))
  col <- rep(largs$col, length(uoid))
  dNA <- data_frame(x_ = NA_real_, y_ = NA_real_)

  largs$y <- NULL
  
  ## do all of this upfront, add the NAs in loop
  a1 <- 
    x$o %>% dplyr::select(object_) %>% 
    inner_join(x$b, 'object_') %>% 
    inner_join(x$bXv, 'branch_') %>% 
    inner_join(x$v, "vertex_") %>% 
    select(x_, y_, branch_, object_)  
  #%>% ungroup ## ungroup to speed up filter
  
  for (i in seq(length(uoid))) {
    largs$col <- col[i]
    a <- a1 %>% filter(object_ == uoid[i])
  if (length(unique(a$branch_)) > 1) {
      a <- do.call(bind_rows, lapply(split(a[, c("x_", "y_")], a$branch_), function(x) bind_rows(x, dNA)))
     # a <- a[-nrow(a), ]
      largs$x <- head(a$x_, -1)
      largs$y <- head(a$y_, -1)
    } else {
    largs$x <- a$x_
    largs$y <- a$y_
    }
    if (type == "pp") {
      do.call(polypath, largs)
    }
    if (type == "l") {
      do.call(lines, largs)
    }
    if (type == "p")
      do.call(points, largs)
  }
  invisible(x)
}


#' @export
#' @rdname gris
as.gris <- function(x, ...)
  UseMethod("as.gris")

#' @export
as.gris.gris <- function(x, ...)
  x

#' as.gris
#' @param x triangulation (package RTriangle)
#' @param type mesh, triangle polygons, or triangle edges
#' @export
#' @importFrom dplyr data_frame
as.gris.triangulation <- function(x, ...) {
  largs <- list(...)
  types = c("mesh", "poly", "line")
  type <- grep(largs$type, types, value = TRUE)
  .tri2gris(x, type = type)
}


.tri2gris <- function(xx, type) {
  o <-
    list(v = data_frame(
      x = xx$P[,1], y = xx$P[,2], vertex_ = seq(nrow(xx$P))
    ))
  
  multi <- !type == "mesh"
  if (type == "line") {
    prims <- xx$E
    primNVerts <- 2
    #stop("gris line creation not yet implemented")
  } else {
    prims <- xx$T
    primNVerts <- 3
  }
  o$b <- data_frame(branch_ = seq(nrow(prims)))
  
  
  if (multi) {
    #prims <- xx$T
    o$b$object_ <- seq(nrow(o$b))
    o$o <- data_frame(object_ = seq(nrow(o$b)))
  } else {
    o$b$object_ <- rep(1, nrow(o$b))
    o$o <- data_frame(object_ = 1)
  }
  
  o$bXv <-
#    data_frame(vertex_ = as.vector(t(prims)), branch_ = rep(seq(nrow(prims)), each = primNVerts))
  tibble(vertex_ = as.vector(t(xx$T)), branch_ = rep(seq(nrow(xx$T)), each = 3))

 class(o) <- c("gris", "list")
  o
}



#' Convert Spatial*DataFrame objects to gris
#'
#' All availables Spatial*DataFrame types are supported, and are interpreted via the Branch model. 
#' The gris function can ingest \code{\link[sp]{SpatialPolygonsDataFrame}}, \code{\link[sp]{SpatialLinesDataFrame}}, \code{\link[sp]{SpatialPointsDataFrame}}, and \code{\link[sp]{SpatialMultiPointsDataFrame}} objects. 
#' 
#' See  \code{vignette("branch-vs-primitives")}
#' @param x Spatial* object
#' @param ... not used
#' @aliases gris.Spatial
#' @return gris
#' @export
#' @rdname sp2gris
gris.Spatial <- function(x, ...) {
  ## one method for all sp (need to dummify if doesn't have a dataframe)
  x <- bld3(x, ...)
  class(x) <- c("gris", "list")
  x
}

gris.SpatialPolygonsDataFrame <- gris.Spatial
gris.SpatialLinesDataFrame <- gris.Spatial
gris.SpatialMultiPointsDataFrame <- gris.Spatial
gris.SpatialPointsDataFrame <- gris.Spatial



#' Not yet implemented
#'
#' @param x 
#'
#' @return gris
#' @export
#'
#' @examples
#' \dontrun{
#'  data(cmu)
#'  
#'  
#'  sp <- as.SpatialPolygonsDataFrame(cmu)
#'  gm <- raster::geom(sp)
#' }
grisFromFortify <- function(x) {
  stop("not yet implemented")
}

#' Not yet implemented
#'
#' @param x 
#'
#' @return gris
#' @export
#'
#' @examples
grisFromRasterGeom <- function(x) {
  stop("not yet implemented")
}

topotype <- function(x) {
  "basic gris"
}


#' @importFrom tibble as_tibble
bld3 <- function(x, ...) {
  tabmap <- spbabel::sptable(x)
  tabdat <- tibble::as_tibble(x)
  tabdat$object_ <- seq(nrow(tabdat))
  grasp(tabdat, tabmap)
}

#' convert from spbabel sptable to gris
#' @importFrom dplyr %>% bind_rows distinct_ mutate select select_
#' @importFrom tibble tibble
#' @importFrom spbabel sptable
grasp <- function(dat1, map1) {
  ## we expect that these attributes, taken together are "unique vertices" potentially shared by neighbours
  v_atts <- c("x_", "y_")
  o_atts <- setdiff(names(map1), v_atts)
  b_atts <- setdiff(o_atts, c("order_", "vertex_"))
  bxv_atts <- c(setdiff(names(map1), c("object_", "island_", v_atts)), "vertex_")
  ## classify unique vertices by unique index
  map1 <- map1 %>%
    mutate(vertex_  = as.integer(factor(do.call(paste, select_(map1, .dots = v_atts)))))
  ## branches, owner object and island status
 b <- map1 %>% distinct_(.dots = b_atts) 
  #b <- map1 %>% dplyr::select(-object_, -island_)
  
  ## four tables (dat1, map2, map4, map5)
  bXv <- map1 %>% dplyr::select_(.dots = bxv_atts)
  v <- map1 %>% distinct_(.dots = c(v_atts, "vertex_"))

  list(o = dat1, b = b, bXv = bXv, v = v)
  
}




.georeference <- function(proj4 = "NA_character_", ...) {
  gg <- list(proj4 = proj4)
  class(gg) <- c("georef", "list")
  gg
}
